src/bluetooth.rs

```diff
@@ -6,187 +6,210 @@
+            _ => Err(Box::from(NOT_SUPPORTED_ERROR)),
+        }
+    };
+
+    ($enum_value: expr, $enum_type: ident, $function_name: ident) => {
+        match $enum_value {
+            &$enum_type::Mock(ref fake) => fake.$function_name(),
+            _ => Err(Box::from(NOT_SUPPORTED_ERROR)),
+        }
+    };
 }
 
 impl BluetoothAdapter {
     #[cfg(all(target_os = "linux", feature = "bluetooth"))]
     pub fn init() -> Result<BluetoothAdapter, Box<Error>> {
         let bluez_adapter = try!(BluetoothAdapterBluez::init());
-        Ok(BluetoothAdapter {adapter: Arc::new(bluez_adapter)})
+        Ok(BluetoothAdapter::Bluez(Arc::new(bluez_adapter)))
     }
 
     #[cfg(all(target_os = "android", feature = "bluetooth"))]
     pub fn init() -> Result<BluetoothAdapter, Box<Error>> {
         let blurdroid_adapter = try!(BluetoothAdapterAndroid::get_adapter());
-        Ok(BluetoothAdapter {adapter: Arc::new(blurdroid_adapter)})
+        Ok(BluetoothAdapter::Android(Arc::new(blurdroid_adapter)))
     }
 
-    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"), feature = "bluetooth-test")))]
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"))))]
     pub fn init() -> Result<BluetoothAdapter, Box<Error>> {
         let adapter = try!(BluetoothAdapterEmpty::init());
-        Ok(BluetoothAdapter {adapter: Arc::new(adapter)})
+        Ok(BluetoothAdapter::Empty(Arc::new(adapter)))
     }
 
     #[cfg(feature = "bluetooth-test")]
-    pub fn init() -> Result<BluetoothAdapter, Box<Error>> {
-        Ok(BluetoothAdapter{adapter: FakeBluetoothAdapter::new_empty()})
-    }
-
-    #[cfg(all(target_os = "linux", feature = "bluetooth"))]
-    fn get_adapter(&self) -> Arc<BluetoothAdapterBluez> {
-        self.adapter.clone()
-    }
-
-    #[cfg(all(target_os = "android", feature = "bluetooth"))]
-    fn get_adapter(&self) -> Arc<BluetoothAdapterAndroid> {
-        self.adapter.clone()
-    }
-
-    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"), feature = "bluetooth-test")))]
-    fn get_adapter(&self) -> Arc<BluetoothAdapterEmpty> {
-        self.adapter.clone()
-    }
-
-    #[cfg(feature = "bluetooth-test")]
-    fn get_adapter(&self) -> Arc<FakeBluetoothAdapter> {
-        self.adapter.clone()
+    pub fn init_mock() -> Result<BluetoothAdapter, Box<Error>> {
+        Ok(BluetoothAdapter::Mock(FakeBluetoothAdapter::new_empty()))
     }
 
     pub fn get_id(&self) -> String {
-        self.get_adapter().get_id()
+        get_inner_and_call!(self, BluetoothAdapter, get_id)
     }
 
     #[cfg(feature = "bluetooth-test")]
     pub fn set_id(&self, id: String) {
-        self.get_adapter().set_id(id)
+        match self {
+            &BluetoothAdapter::Mock(ref fake_adapter) => fake_adapter.set_id(id),
```

Comment:
To remain consistent, we should use `get_inner_and_call_test_func` here as well.